commit 76845de2188bd1dfca8ec9590132adb96ae5785c (refs/changes/27/1527/2)
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2017-02-24T10:17:12-08:00 (2 years, 7 months ago)
    
    AGENT-1062 vm-agent should use cueball
    Reviewed by: Orlando Vazquez <orlando@joyent.com>
    Approved by: Orlando Vazquez <orlando@joyent.com>

diff --git a/lib/vm-agent.js b/lib/vm-agent.js
index 4aa4c4a..5dc56f6 100644
--- a/lib/vm-agent.js
+++ b/lib/vm-agent.js
@@ -118,6 +118,7 @@ var fs = require('fs');
 var path = require('path');
 
 var assert = require('assert-plus');
+var cueball = require('cueball');
 var diff = require('deep-diff').diff;
 var vasync = require('vasync');
 var vmadm = require('vmadm');
@@ -139,6 +140,7 @@ var MAX_UPDATE_DELAY_MS = 30000;
 
 function VmAgent(options) {
     var self = this;
+    var agent;
     var packageJson = path.join(path.dirname(__dirname), 'package.json');
     var userAgent;
 
@@ -148,6 +150,7 @@ function VmAgent(options) {
         'options.periodic_interval');
     assert.uuid(options.server_uuid, 'options.server_uuid');
     assert.string(options.vmapi_url, 'options.vmapi_url');
+    assert.optionalObject(options.cueballHttpAgent, 'options.cueballHttpAgent');
 
     // Load the list of fields that VmWatcher will be watching, which we'll
     // use when comparing objects. Since VMAPI and vmadm have different default
@@ -171,13 +174,18 @@ function VmAgent(options) {
 
     assert(self.version, 'missing package.json version');
 
+    if (options.cueballHttpAgent) {
+        agent = new cueball.HttpAgent(options.cueballHttpAgent);
+    }
+
     userAgent = 'vm-agent/' + self.version
         + ' (node/' + process.versions.node + ')'
         + ' server/' + self.server_uuid;
 
     self.vmapiClient = new VMAPI({
-        url: options.vmapi_url,
+        agent: agent,
         log: options.log,
+        url: options.vmapi_url,
         userAgent: userAgent
     });
 
diff --git a/package.json b/package.json
index 9491510..63de36a 100644
--- a/package.json
+++ b/package.json
@@ -1,12 +1,13 @@
 {
     "name": "vm-agent",
     "description": "SmartDC Node VM Agent",
-    "version": "1.6.0",
+    "version": "1.6.1",
     "author": "Joyent (joyent.com)",
     "private": true,
     "dependencies": {
         "assert-plus": "0.2.0",
         "bunyan": "1.5.1",
+        "cueball": "2.1.1",
         "deep-diff": "0.3.3",
         "eslint": "1.10.1",
         "kthxbai": "0.4.0",
diff --git a/sapi_manifests/vm-agent/template b/sapi_manifests/vm-agent/template
index b50a21c..5e1db68 100644
--- a/sapi_manifests/vm-agent/template
+++ b/sapi_manifests/vm-agent/template
@@ -1,4 +1,21 @@
 {
+    "cueballHttpAgent": {
+        "initialDomains": [
+            "{{{vmapi_domain}}}"
+        ],
+        "maximum": 100,
+        "recovery": {
+            "default": {
+                "delay": 250,
+                "maxDelay": 1000,
+                "maxTimeout": 8000,
+                "retries": 5,
+                "timeout": 2000
+            }
+        },
+        "resolvers": ["{{{BINDER_SERVICE}}}"],
+        "spares": 4
+    },
     {{#periodic_interval}}"periodic_interval": {{{periodic_interval}}},{{/periodic_interval}}
     "vmapi_url": "http://{{{VMAPI_SERVICE}}}",
     "no_rabbit": {{#no_rabbit}}true{{/no_rabbit}}{{^no_rabbit}}false{{/no_rabbit}}
